### PR TITLE
Refactor Solver class and rewrite ConvTasNet/train.py accordingly

### PR DESCRIPTION
**Pull request by mpariente**
Refactor `Solver` to enable easy overwriting of sub-methods and give more freedom to the user.

### CODE REVIEW

File: asteroid/engine/solver.py [hunk 1]
```diff
@@ -52,37 +66,38 @@ def __init__(self, data, model, loss_class, optimizer, training_conf):
...
             start = time.time()
-            tr_avg_loss = self._run_one_epoch(epoch)
+            tr_avg_loss = self.run_one_epoch(epoch)
             print('-' * 85)
             print('Train Summary | End of Epoch {0} | Time {1:.2f}s | '
                   'Train Loss {2:.3f}'.format(epoch + 1, time.time() - start,
                                               tr_avg_loss))
             print('-' * 85)
-            if self.checkpoint:
+            if self.tr_conf['checkpoint']:
```
Comment (sunits) [inline asteroid/engine/solver.py]: Should we use get function here? Just to be consistent

File: asteroid/engine/solver.py [hunk 2]
```diff
@@ -95,69 +110,92 @@ def train(self):
...
                 self.early_val_no_impv = 0
             # Adjust learning rate (halving)
-            if self.half_lr and self.val_no_impv >= 3:
+            if self.tr_conf['half_lr'] and self.val_no_impv >= 3:
                 self.multiply_learning_rate(self.optimizer, 0.5)
                 self.ep_half_lr.append(epoch)
                 self.val_no_impv = 0
             # Early stopping
-            if self.early_val_no_impv >= 8 and self.early_stop:
+            if self.early_val_no_impv >= 8 and self.tr_conf['early_stop']:
```
Comment (sunits) [inline asteroid/engine/solver.py]: Remove constants. Put it in default dictionary.

Comment (sunits) [reply asteroid/engine/solver.py]: Same for 1.001 in line 106

File: asteroid/engine/solver.py [hunk 3]
```diff
@@ -95,69 +110,92 @@ def train(self):
...
             self.tr_loss[epoch] = tr_avg_loss
             self.cv_loss[epoch] = val_loss
             if val_loss < self.best_val_loss:
                 self.best_val_loss = val_loss
-                file_path = os.path.join(self.save_folder, self.model_path)
-                self.save_model(file_path, epoch)
+                self.save_model(self.model_path, epoch)
 
-    def _run_one_epoch(self, epoch, validation=False):
+    def run_one_epoch(self, epoch, validation=False):
```
Comment (sunits) [inline asteroid/engine/solver.py]: Let us discuss about the validation flag.

File: asteroid/engine/solver.py [hunk 4]
```diff
@@ -95,69 +110,92 @@ def train(self):
...
+
+    def compute_loss(self, targets, est_targets, infos=None):
+        """ Computes the loss to be backpropred.
+        `infos` is a dictionary that is passed to the loss function.
+        """
+        return self.loss_class.compute(targets, est_targets, infos=infos)
+
+    def loss_backward_and_step(self, loss, validation=False):
+        """ Backpropagation of the loss and gradient descend step."""
+        if not validation and not torch.isnan(loss):
```
Comment (sunits) [inline asteroid/engine/solver.py]: What happens if loss is NaN? May be do that check in compute_loss. 
How about making the some of these functions private? Should they be accessible outside the class?

File: asteroid/engine/solver.py (hunk 1 above)
Comment (mpariente) [reply asteroid/engine/solver.py]: `get` is called when the key might not be here. In this case, the key is here.

File: asteroid/engine/solver.py (hunk 2 above)
Comment (mpariente) [reply asteroid/engine/solver.py]: You're right, thanks. Callbacks will make this so much better

File: asteroid/engine/solver.py (hunk 4 above)
Comment (mpariente) [reply asteroid/engine/solver.py]: If loss is NaN, optimizer.step() makes all the weights NaN, and the network is unrecoverable.
The good approach will be to find the reason and prevent it. See [the anomaly detection tool from pytorch](https://pytorch.org/docs/master/autograd.html#anomaly-detection). I need to take time to set it up.

Comment (sunits) [reply asteroid/engine/solver.py]: Put a Warning  for now.
